Be more like rioxarray in CF attribute handling

Kirill888 opened the pull request:
netcdf/cf related changed:

- For CRSs supported by netdcf/cf convention include all attributes as reported by `pyproj.to_cf()` method.
- Look in `crs_wkt` attribute as well as in `spatial_ref` when looking for and interpreting CRS coordinate.


Some further API tweaking
- `with_crs="name"|True|False` argument to `xr_coord` was confusing, so it was replaced with
- `crs_coord_name="name"|None` where `None` is the same `False` was before, i.e. do not create a separate CRS coordinate.
- use default WKT format in `.wkt` property of the CRS object

Kirill888 committed:
Attach all CF attributes to CRS coordinate

- be more like rioxarray
- understand `crs_wkt` from CF convention

Kirill888 committed:
Use default WKT version in `CRS.wkt` property

Kirill888 committed:
Refactor xr_coord

Change `with_crs: str|bool` to be `crs_coord_name: str|None`

gadomski commented on odc/geo/_xr_interop.py:
```diff
@@ -123,21 +123,16 @@ def spatial_dims(
-        grid_mapping_name = "latitude_longitude"
-
+    cf = crs.proj.to_cf()
+    cf.setdefault("grid_mapping_name", "??")
```
Is there a reason `??` is used? Should the be a constant somewhere? Or is this just for printing purposes?

Kirill888 replied:
Actually I'm not sure we should be setting it to anything. Basically `grid_mapping_name` is an `enum` defined by CF convention, so if your CRS is not supported by CF convention, it should probably not have `grid_mapping_name` defined at all. We are relying on it being there do detect "CRS coordinates" attached to an xarray. But this should probably change to look for any of these attributes instead: `grid_mapping_name, spatial_ref, crs_wkt`.

Not sure what that means for compatibility of files produced with `.to_netcdf(...)`. 


https://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/apf.html

Kirill888 replied:
hm, looks like we are not even using that attribute for locating crs coordinate. I think it should just be left unset if it's not defined in the output of `.to_cf()`.

I should also add tests for round-tripping CRS to netcdf/zarr/tiff. it could be that non-standard name is worse than no name.

gadomski commented on odc/geo/_xr_interop.py:
```diff
@@ -123,21 +123,16 @@ def spatial_dims(
+    cf = crs.proj.to_cf()
+    cf.setdefault("grid_mapping_name", "??")
     epsg = 0 if crs.epsg is None else crs.epsg
+    crs_wkt = cf.get("crs_wkt", None) or crs.wkt
```
```suggestion
    crs_wkt = cf.get("crs_wkt", crs.wkt)
```

Kirill888 replied:
Reason why I used `or` syntax is because it short-circuits, so we don't need to recompute `wkt` string if it is already included in the cf convention output (it always is, as far as I can tell). It's probably not a very expensive computation, but there is also no telling how often this code might be used in an application.

Kirill888 committed:
Do not set grid_mapping_name when not defined

Previously we were using "??" as a fallback, but `grid_mapping_name`
attribute is a kind of an enum defined by CF convention, so best not to
inject things not define in that spec.
